ambry-network/src/main/java/com.github.ambry.network/Selector.java

```diff
@@ -291,13 +301,17 @@ public void poll(long timeoutMs) throws IOException {
    * completed I/O.
    *
    * @param timeoutMs The amount of time to wait, in milliseconds. If negative, wait indefinitely.
-   * @param sends The list of new sends to begin
+   * @param sends The list of new sends to initiate.
    *
    * @throws IOException If a send is given for which we have no existing connection or for which there is
    *         already an in-progress send
    */
   @Override
   public void poll(long timeoutMs, List<NetworkSend> sends) throws IOException {
+    if (executorPool != null) {
+      pollWithExecutorPool(timeoutMs, sends);
+      return;
```

Comment:
minor: move the code below line 314 into another method like `pollOnMainThread()` and use an if-else to switch between them.